analyzer/src/main/kotlin/managers/utils/NuGetInspector.kt

```diff
@@ -0,0 +1,324 @@
+        buildString {
+            party.name?.let { append(it) }
+            party.email?.let {
+                append(if (party.name != null && it.isNotBlank()) " <$it>" else it)
+            }
+        }.takeIf { it.isNotBlank() }
+    }
+
+private fun NuGetInspector.PackageData.toPackageReference() =
+    PackageReference(
+        id = Identifier(type = TYPE, namespace = "", name = name, version = version),
+        dependencies = dependencies.mapTo(sortedSetOf()) {
+            PackageReference(
+                id = Identifier(type = TYPE, namespace = "", name = it.name, version = it.version),
+            )
+        }
+    )
+
+internal fun NuGetInspector.Result.toOrtProject(
+    managerName: String,
+    analysisRoot: File,
+    definitionFile: File,
+): Project {
+    val spec = resolveLocalSpec(definitionFile)?.let { NuGetSupport.XML_MAPPER.readValue<PackageSpec>(it) }
+    val id = Identifier(
+        type = managerName,
+        namespace = "",
+        name = spec?.metadata?.id ?: definitionFile.relativeTo(analysisRoot).invariantSeparatorsPath,
+        version = spec?.metadata?.version.orEmpty()
+    )
+    val workingDir = definitionFile.parentFile
+
+    val nestedPackages: MutableList<NuGetInspector.PackageData> = mutableListOf()
+
+    packages.forEach {
+        it.packages.forEach { pkg -> nestedPackages.add(pkg) }
+    }
+
+    val packageReferences = nestedPackages.mapTo(sortedSetOf()) {
+        it.toPackageReference()
+    }
+
+    val scopes = sortedSetOf(Scope(headers.first().projectFramework, packageReferences))
+    return Project(
+        id = id,
+        definitionFilePath = VersionControlSystem.getPathInfo(definitionFile).path,
+        vcs = VcsInfo.EMPTY,
+        authors = parseAuthors(spec),
+        vcsProcessed = PackageManager.processProjectVcs(workingDir),
+        declaredLicenses = parseLicenses(spec),
+        homepageUrl = "",
+        scopeDependencies = scopes
+    )
+}
+
+internal fun Set<NuGetInspector.PackageData>.toOrtPackages(): Set<Package> =
+    groupBy { "${it.name}:${it.version}" }.mapTo(mutableSetOf()) { (_, packages) ->
+        val pkg = packages.first()
+
+        fun NuGetInspector.PackageData.getHash(): Hash = Hash.create(
+            sha512?.lowercase() ?: sha256?.lowercase() ?: sha1?.lowercase() ?: md5?.lowercase() ?: ""
+        )
+
+        val id = Identifier(type = TYPE, namespace = pkg.namespace.orEmpty(), name = pkg.name, version = pkg.version)
+        val declaredLicenses: SortedSet<String> = sortedSetOf()
+        val pkgDeclaredLicense = pkg.declaredLicense.orEmpty()
+
+        if (pkgDeclaredLicense.isNotBlank()) {
+            val licenseData = yamlMapper.readValue<Map<String, String>>(pkgDeclaredLicense)
+            licenseData["LicenseUrl"]?.also { declaredLicenses += it }
+            licenseData["License"]?.also { declaredLicenses += it }
+            licenseData["LicenseExpression"]?.also { declaredLicenses += it }
```

Comment:
If none of the values are null, all are added to the `declaredLicenses`.
Which will result in a declared license like: `MIT AND MIT AND https://www.nuget.org/packages/Newtonsoft.Json/12.0.2/license` due to the implicit `AND` in `DeclaredLicenseProcessor.process(declaredLicenses)`.

However, `DeclaredLicenseProcessor.process(declaredLicenses)` also fails to parse the URL as a declared license.

In Maven it is done differently:
https://github.com/oss-review-toolkit/ort/blob/94231346848a0b15fb461727b5c30c4f598ef9b5/analyzer/src/main/kotlin/managers/utils/MavenSupport.kt#L148-L151
The URL and the `comment` is only taken as a fallback if the `name` is null or empty. I think we should use a similar approach here.

Reply:
The way licenses are now provided by NuGet has changed. I am a bit at loss to create something entirely new to support this.

Reply:
The NuGet license documentation practice is evolving and we legacy but common and new ways.

We have three field: license URL, type and expression.
The field named "License" as returned by nuget-inspector is redundatnt and will be removed in the next release.


- The new way is to return a license expression field that can either be of type "file" or type "expression"
  - If the type is "expresison", this should an SPDX license expression. With one twist: it can be "UNLICENSED" (ignoring case) as per https://github.com/linksplatform/Data.Sequences/blob/970664ef92c6b932f9c0c07e0ec6ba52ec792720/csharp/Platform.Data.Sequences/Platform.Data.Sequences.csproj#L15 in which case I suggest we return "LicenseRef-scancode-unknown"

  - If the type is "file", this point to a license file path relative to the root of the package.

- The legacy but common way is to get a license URL that points to various things and is unreliable.


NuGet says that if a license expression is present the URL should be ignored.
In particular NuGet will generate a URL based on the expression such as https://licenses.nuget.org/(LGPL-2.0-only%20WITH%20FLTK-exception) and this is redundant with the expression.


So the proposed approach is this:

- If there is an expression of type "expression" use this as-is as an SPDX expression
- If there is no expression and but there is a URL, this needs to be mapped  and for this we can reuse the mappings that exist.
- If there is an expression of type "file", there is nothing we can do at this stage and this should be handled in a future enhancement to process the file correctly (though note that the file would not be present for a dependency)
  - If there is a URL and a file, we apply the URL mapping as for a bare URL.

Reply:
We implemented the fall back from expression to URL, so I am marking this as resolved.